cpp/src/Ice/ConnectionI.cpp

```diff
@@ -2807,6 +2765,41 @@ Ice::ConnectionI::sendMessage(OutgoingMessage& message)
     assert(_state >= StateActive);
     assert(_state < StateClosed);
 
+    bool isHeartbeat = static_cast<uint8_t>(message.stream->b[8]) == IceInternal::validateConnectionMsg;
+    if (!isHeartbeat)
+    {
+        cancelInactivityTimerTask();
+    }
+
+    // If we're sending a heartbeat, there is a chance the connection is inactive and that we need to schedule the
+    // inactivity timer task.
+    // It's ok to do this before actually sending the heartbeat since the heartbeat does not count as an "activity".
+    if (isHeartbeat && _inactivityTimerTask && // null when the inactivity timeout is infinite
+        !_inactivityTimerTaskScheduled &&      // we never reschedule this task
+        _state == StateActive &&               // only schedule the task if the connection is active
+        _dispatchCount == 0 &&                 // no pending dispatch
+        _asyncRequests.empty())                // no pending invocation
+
+    {
+        bool isInactive = true;
+
+        // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the inactivity
```

Comment:
We can't just check if _sendStreams is empty and ignore the situation where _sendStreams holds heartbeats.

If we're back-pressured for a long while by a healthy peer, it will keep sending us heartbeats and there is no reason for the inactivity timeout to _not_ kick-in in this situation.

Question: when we cancel an invocation (typically because it timed out), do we check the _sendStreams to see if  it's still there & remove it?

Reply:
> Question 1: when we cancel an invocation (typically because it timed out), do we check the _sendStreams to see if it's still there & remove it?

Yes, see the `ConnectionI::asyncRequestCanceled` method implementation.
 
> Question 2: should we try to optimize the heartbeat queuing situation? Like not enqueue heartbeats in _sendStreams? I think it could get confusing, especially since heartbeats are tiny messages sent only when we don't send anything else.

It would indeed be confusing.